src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp

```diff
@@ -0,0 +1,203 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Tags.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/AnalyticData/AnalyticData.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/RegularSphericalWave.hpp"
+#include "PointwiseFunctions/GeneralRelativity/TagsDeclarations.hpp"
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \cond
+namespace PUP {
+class er;  // IWYU pragma: keep
+}  // namespace PUP
+namespace Tags {
+template <typename Tag>
+struct dt;
+}  // namespace Tags
+/// \endcond
+
+// IWYU pragma: no_include <pup.h>
+
+namespace CurvedScalarWave {
+namespace AnalyticData {
+/*!
+ * \brief Analytic initial data for scalar waves in \f$3+1\f$D curved spacetime
+ *
+ * \details When evolving a scalar field propagating through curved spacetime,
+ * this class provides methods to initialize the scalar-field and spacetime
+ * variables using analytic solution(s) of the flat-space scalar-wave equation
+ * and of Einstein equations. That the coordinate profile of the scalar
+ * field \f$\Psi\f$ in curved spacetime is the same as that in flat spacetime
+ * is our primary identification, allowing it to be initialized using any
+ * member class of `ScalarWave::Solutions`. Solving constraints of the
+ * `CurvedScalarWave` system, we also identify the coordinate spatial derivative
+ * of the scalar field \f$\Phi_i\f$ in curved spacetime with the same in flat
+ * spacetime. The definition of \f$\Pi\f$ comes from requiring it to be the
+ * future-directed time derivative of the scalar field:
+ *
+ * \f{align}
+ * \Pi := -n^a \partial_a \Psi
+ *     =  \frac{1}{\alpha}\left(\beta^k \Phi_k - {\partial_t\Psi}\right),
+ * \f}
+ *
+ * where \f$n^a\f$ is the unit normal to spatial slices of the spacetime
+ * foliation.
+ */
+template <typename ScalarFieldData, typename BackgroundGrData>
+class ScalarWaveGr : public MarkAsAnalyticData {
+ public:
+  static constexpr size_t volume_dim = 3;
```

Comment:
why restrict to 3?  Just make sure the two template arguments have the same `volume_dim` and use that

Reply:
Updated